server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java

```diff
@@ -0,0 +1,442 @@
+                        exp
+                    );
+                }
+
+                @Override
+                public String executor() {
+                    return ThreadPool.Names.SAME;
+                }
+
+                @Override
+                public AddVotingConfigExclusionsResponse read(StreamInput in) throws IOException {
+                    return new AddVotingConfigExclusionsResponse(in);
+                }
+            }
+        );
+    }
+
+    private void clearVotingConfigAfterSuccessfulDecommission() {
+        final ClearVotingConfigExclusionsRequest clearVotingConfigExclusionsRequest = new ClearVotingConfigExclusionsRequest();
+        clearVotingConfigExclusionsRequest.setWaitForRemoval(true);
+        transportService.sendRequest(
+            transportService.getLocalNode(),
+            ClearVotingConfigExclusionsAction.NAME,
+            clearVotingConfigExclusionsRequest,
+            new TransportResponseHandler<ClearVotingConfigExclusionsResponse>() {
+                @Override
+                public void handleResponse(ClearVotingConfigExclusionsResponse response) {
+                    logger.info("successfully cleared voting config after decommissioning");
+                }
+
+                @Override
+                public void handleException(TransportException exp) {
+                    logger.debug(new ParameterizedMessage("failure in clearing voting config exclusion after decommissioning"), exp);
+                }
+
+                @Override
+                public String executor() {
+                    return ThreadPool.Names.SAME;
+                }
+
+                @Override
+                public ClearVotingConfigExclusionsResponse read(StreamInput in) throws IOException {
+                    return new ClearVotingConfigExclusionsResponse(in);
+                }
+            }
+        );
+    }
+
+    /**
+     * Registers new decommissioned attribute metadata in the cluster state with {@link DecommissionStatus#DECOMMISSION_INIT}
+     * <p>
+     * This method can be only called on the cluster-manager node. It tries to create a new decommissioned attribute on the master
+     * and if it was successful it adds new decommissioned attribute to cluster metadata.
+     * <p>
+     * This method ensures that request is performed only on eligible cluster manager node
+     *
+     * @param decommissionAttribute register decommission attribute in the metadata request
+     * @param listener              register decommission listener
+     */
+    private void registerDecommissionAttribute(
+        final DecommissionAttribute decommissionAttribute,
+        final ActionListener<ClusterStateUpdateResponse> listener
+    ) {
+        if (!transportService.getLocalNode().isClusterManagerNode()
+            || nodeHasDecommissionedAttribute(transportService.getLocalNode(), decommissionAttribute)) {
```

Comment:
we don't need this . This will  be executed only on `ClusterManager` .  Even after abdication, it will go to elected cluster manager.

Reply:
pls verify the `add voting config` is idempotent as it will be retried on new cluster manager.

Reply:
> we don't need this . This will be executed only on ClusterManager . Even after abdication, it will go to elected cluster manager.

There could be a case where the request to exclude is acknowledged but the leader is not yet abdicated (slow clusters with tasks piled up etc) and in that case we never want to initiate any kind of decommissioning to happen on a `to-be-decommissioned` leader. This check ensures that operation is always on active leader which will never be decommissioned

Reply:
> pls verify the add voting config is idempotent as it will be retried on new cluster manager.

added checks so that we don't exclude same nodes multiple times